Microsoft.DotNet.Interactive/PackageRestoreContext.cs

```diff
@@ -35,78 +49,98 @@ public PackageRestoreContext()
 
         public string Name { get; }
 
-        public Task<ResolvedNugetPackageReference> GetResolvedNugetPackageReference(string packageName)
+        public async Task<ResolvedPackageReference> GetResolvedPackageReference(string packageName)
         {
-            var references = GetResolvedNugetReferences();
+            await Restore();
 
-            return Task.FromResult(references[packageName]);
+            var references = GetResolvedReferences();
+            return references[packageName];
         }
 
-        public async Task<AddNugetResult> AddPackage(
+        public bool AddPackagReference(
             string packageName,
             string packageVersion = null,
-            string restoreSources = null)
+            string restoreSources = null,
+            string sourceCode = null)
         {
-            var requestedPackage = new NugetPackageReference(packageName, packageVersion, restoreSources);
+            var requestedPackage = new PackageReference(packageName, packageVersion, restoreSources, sourceCode);
+            var packageKey = requestedPackage.PackageKey;
 
-            if (!string.IsNullOrEmpty(packageName) && _nugetPackageReferences.TryGetValue(requestedPackage, out _))
+            lock (_packageReferences)
```

Comment:
Rather than needing a `lock`, you can use `ConcurrentDictionary`.

Reply:
Agreed